lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs

```diff
@@ -380,6 +370,32 @@ txWitnessSize = \case
 txWitnessTagSize :: Int
 txWitnessTagSize = 1
 
+-- | Construct the pre-cursor to a @TxWitness@. For internal use only.
+constructGenericWitnessSignature
+    :: Hash "sigData"
+    -> Hash "Genesis"
+    -> (ByteString -> ByteString)
+    -> ByteString
+constructGenericWitnessSignature (Hash txHash) (Hash block0Hash) sign =
+        sign (block0Hash <> txHash)
```

Comment:
I wanted to avoid the situation where `block0Hash <> txHash` was changed and we forgot to change both instances.

The `Hash tag` type does not protect us after we've unwrapped it, and I thought just changing the order of the arguments could potentially go un-noticed.

I definitely agree on that there's *too much* indirection though. Will try to make `utxoWitness` and `legacyUtxoWitness` redundant, and see how inlining looks.

Reply:
When I am tempted to created extra indirection and abstraction in this form, I usually think twice and remember: 

https://github.com/input-output-hk/cardano-sl/blob/2817a37d95a71ec8380d82543d24dd94a22472a3/client/src/Pos/Client/Txp/Util.hs#L632

```
- createTx
    |--> createGenericTxSingle
        |--> createGenericTx
            |--> runTxCreator
                |--> prepareInpsOuts
                    |--> prepareTxWithFee
                        |-->  ....
```           

If you remove the noise and cut out the indirections, this boils down to ~5 lines of code. And all of this for what? Really, only two types of transactions (redeem vs normal). 
I'd rather maintain two 5-line straightforward functions with a bit of duplication than having to go over the indirection hell that can be created as a result from this. 

If the duplication is something that bother you, then instead of `constructGenericWitnessSignature`, what about simply introducing: 

```hs
mkWitnessPayload :: Hash "Genesis" -> Hash "sigData" -> ByteString
mkWitnessPayload (Hash genesis) (Hash sigData) = genesis <> sigData
```

But to be honest, I don't think it saves us much in this case. So, we better have unit tests checking for regression in the format in case we update and forget to update one.